metis-media-service/src/main/java/eu/europeana/metis/mediaprocessing/RdfDeserializerImpl.java

```diff
@@ -45,41 +44,65 @@
  */
 class RdfDeserializerImpl implements RdfDeserializer {
 
-  private final IUnmarshallingContext context;
-
-  private final XPathExpression getObjectExpression;
-  private final XPathExpression getHasViewExpression;
-  private final XPathExpression getIsShownAtExpression;
-  private final XPathExpression getIsShownByExpression;
-
-  /**
-   * Constructor.
-   *
-   * @throws RdfConverterException In case something went wrong constructing this object.
-   */
-  RdfDeserializerImpl() throws RdfConverterException {
+  private final UnmarshallingContextWrapper unmarshallingContext = new UnmarshallingContextWrapper();
+
+  private final XPathExpressionWrapper getObjectExpression = new XPathExpressionWrapper(
+          xPath -> xPath.compile("/rdf:RDF/ore:Aggregation/edm:object/@rdf:resource"));
+  private final XPathExpressionWrapper getHasViewExpression = new XPathExpressionWrapper(
+          xPath -> xPath.compile("/rdf:RDF/ore:Aggregation/edm:hasView/@rdf:resource"));
+  private final XPathExpressionWrapper getIsShownAtExpression = new XPathExpressionWrapper(
+          xPath -> xPath.compile("/rdf:RDF/ore:Aggregation/edm:isShownAt/@rdf:resource"));
+  private final XPathExpressionWrapper getIsShownByExpression = new XPathExpressionWrapper(
+          xPath -> xPath.compile("/rdf:RDF/ore:Aggregation/edm:isShownBy/@rdf:resource"));
+
+  private static class XPathExpressionWrapper extends
+          ThreadSafeWrapper<XPathExpression, RdfDeserializationException> {
+
+    public XPathExpressionWrapper(
+            ThrowingFunction<XPath, XPathExpression, XPathExpressionException> expressionCreator) {
+      super(() -> {
+        final XPath xPath = XPathFactory.newInstance().newXPath();
```

Comment:
Ok so this is much better.
I checked the issues we had with transformation and they don't exactly apply here, so we don't really have to use Saxon I suppose.
But in this code there is still a synchronization issue.
The call to XPathFactory.newInstance() is not thread safe and in this case it can be accessed from two threads at the same if those two threads are requesting an expression of two different types. So it works for the same type, but not for two different types.
Do you understand what I mean?

Reply:
I understand. I think the documentation is a bit vague, I assumed that what was meant was that the instance thus created was not thread safe (but the static method was). But it is an easy fix.

Reply:
Yes it is quite vague..
I hope we are fine thread wise here though now.